CommonTools/RecoAlgos/plugins/SimClusterRecHitAssocitionProducer.cc

```diff
@@ -0,0 +1,136 @@
+          caloRechitTags_, [this](const edm::InputTag& tag) { return mayConsume<edm::View<CaloRecHit>>(tag); })),
+      scCollectionToken_(consumes<SimClusterCollection>(pset.getParameter<edm::InputTag>("simClusters"))) {
+  for (auto& tag : caloRechitTags_) {
+    const std::string& label = tag.instance();
+    produces<edm::Association<SimClusterCollection>>(label + "ToSimClus");
+  }
+  produces<std::unordered_map<int, float>>();
+}
+
+SimClusterRecHitAssociationProducer::~SimClusterRecHitAssociationProducer() {}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void SimClusterRecHitAssociationProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  auto simClusterToRecEnergy = std::make_unique<std::unordered_map<int, float>>();
+  edm::Handle<SimClusterCollection> scCollection;
+  iEvent.getByToken(scCollectionToken_, scCollection);
+  std::unordered_map<size_t, IdxAndFraction> hitDetIdToIndex;
+  std::unordered_map<size_t, float> hitDetIdToTotalSimFrac;
+
+  for (size_t s = 0; s < scCollection->size(); s++) {
+    const auto& sc = scCollection->at(s);
+    (*simClusterToRecEnergy)[s] = 0.;
+    for (auto& hf : sc.hits_and_fractions()) {
+      auto entry = hitDetIdToIndex.find(hf.first);
+      // Update SimCluster assigment if detId has been found in no other SCs or if
+      // SC has greater fraction of energy in DetId than the SC already found
+      if (entry == hitDetIdToIndex.end()) {
+        hitDetIdToTotalSimFrac[hf.first] = hf.second;
+        hitDetIdToIndex[hf.first] = {s, hf.second};
+      } else {
+        hitDetIdToTotalSimFrac[hf.first] += hf.second;
+        if (entry->second.second < hf.second)
+          hitDetIdToIndex[hf.first] = {s, hf.second};
+      }
+    }
+  }
+
+  std::unordered_map<int, float> hitDetIdToTotalRecEnergy;
+
+  for (size_t i = 0; i < caloRechitCollectionTokens_.size(); i++) {
+    std::string label = caloRechitTags_.at(i).instance();
+    std::vector<size_t> rechitIndices;
+
+    edm::Handle<edm::View<CaloRecHit>> caloRechitCollection;
+    iEvent.getByToken(caloRechitCollectionTokens_.at(i), caloRechitCollection);
+
+    for (size_t h = 0; h < caloRechitCollection->size(); h++) {
+      const CaloRecHit& caloRh = caloRechitCollection->at(h);
+      size_t id = caloRh.detid().rawId();
+      auto entry = hitDetIdToTotalRecEnergy.find(id);
+      float energy = caloRh.energy();
+      if (entry == hitDetIdToTotalRecEnergy.end())
+        hitDetIdToTotalRecEnergy[id] = energy;
+      else
+        hitDetIdToTotalRecEnergy.at(id) += energy;
+
+      int match = hitDetIdToIndex.find(id) == hitDetIdToIndex.end() ? -1 : hitDetIdToIndex.at(id).first;
+      float fraction = match != -1 ? hitDetIdToTotalSimFrac.at(id) : 1.;
+      if (simClusterToRecEnergy->find(match) == simClusterToRecEnergy->end())
+        (*simClusterToRecEnergy)[match] = energy * fraction;
+      else
+        simClusterToRecEnergy->at(match) += energy * fraction;
```

Comment:
isn't the default inserted value going to be zero? In this case the following is the same and is shorter
```suggestion
      (*simClusterToRecEnergy)[match] += energy * fraction;
```